File: shared-code/typescript/telemetry/src/profiler.ts
import Pyroscope from "@pyroscope/nodejs";

export type Profiler = {
  start: () => void;
  shutdown: () => Promise<void>;
};

export function createProfiler(serviceName: string): Profiler {
  Pyroscope.init({
    serverAddress: process.env.PYROSCOPE_URL,
    appName: serviceName,
    flushIntervalMs: 10000,
    wall: {
      samplingDurationMs: 10000,
      samplingIntervalMicros: 1000,
      collectCpuTime: true
    },
    tags: {
      lang: "node"
    }
  });

  return {
    start: () => Pyroscope.start(),
    shutdown: () => Pyroscope.stop()
  };
}

File: shared-code/go/telemetry/profiler.go
package telemetry

import (
	"errors"
	"os"

	"github.com/grafana/pyroscope-go"
)

type Profiler struct {
	serviceName string
	client      *pyroscope.Profiler
}

func NewProfiler(serviceName string) *Profiler {
	return &Profiler{serviceName: serviceName}
}

func (p *Profiler) Start() error {
	if p.client != nil {
		return nil
	}

	serverAddress, ok := os.LookupEnv("PYROSCOPE_URL")
	if !ok || serverAddress == "" {
		return errors.New("Missing PYROSCOPE_URL")
	}

	config := pyroscope.Config{
		ApplicationName: p.serviceName,
		ServerAddress:   serverAddress,
		Tags:            map[string]string{"lang": "go"},
	}

	profiler, err := pyroscope.Start(config)
	if err != nil {
		return err
	}

	p.client = profiler
	return nil
}

func (p *Profiler) Shutdown() error {
	if p.client == nil {
		return nil
	}
	return p.client.Stop()
}
